src/WorkflowFunction.ts

```diff
@@ -0,0 +1,214 @@
+
+/** Constants */
+
+const VALID_PAYLOAD_TYPES = new Set(['function_executed']);
+
+/** Class */
+
+export class WorkflowFunction {
+  /** Function callback_id */
+  public callbackId: string;
+
+  private middleware: WorkflowFunctionMiddleware;
+
+  public constructor(
+    callbackId: string,
+    middleware: WorkflowFunctionMiddleware,
+  ) {
+    validate(callbackId, middleware);
+
+    this.callbackId = callbackId;
+    this.middleware = middleware;
+  }
+
+  public getMiddleware(): Middleware<AnyMiddlewareArgs> {
+    return async (args): Promise<any> => {
+      if (isFunctionEvent(args) && this.matchesConstraints(args)) {
+        return this.processEvent(args);
+      }
+      return args.next();
+    };
+  }
+
+  private matchesConstraints(args: SlackWorkflowFunctionMiddlewareArgs): boolean {
+    return args.payload.function.callback_id === this.callbackId;
+  }
+
+  private async processEvent(args: AllWorkflowFunctionMiddlewareArgs): Promise<void> {
+    const functionArgs = prepareFunctionArgs(args);
+    const stepMiddleware = this.getStepMiddleware();
+    return processStepMiddleware(functionArgs, stepMiddleware);
+  }
+
+  private getStepMiddleware(): WorkflowFunctionMiddleware {
+    return this.middleware;
+  }
+}
+
+/** Helper Functions */
+
+export function validate(callbackId: string, listeners: WorkflowFunctionMiddleware): void {
+  // Ensure callbackId is valid
+  if (typeof callbackId !== 'string') {
+    const errorMsg = 'WorkflowFunction expects a callback_id as the first argument';
+    throw new WorkflowFunctionInitializationError(errorMsg);
+  }
+
+  // Ensure all listeners are functions
+  listeners.forEach((listener) => {
+    if (!(listener instanceof Function)) {
+      const errorMsg = 'All WorkflowFunction listeners must be functions';
+      throw new WorkflowFunctionInitializationError(errorMsg);
+    }
+  });
+}
+
+/**
+ * `processStepMiddleware()` invokes each callback for lifecycle event
+ * @param args workflow_step_edit action
+ */
+export async function processStepMiddleware(
+  args: AllWorkflowFunctionMiddlewareArgs,
+  middleware: WorkflowFunctionMiddleware,
+): Promise<void> {
+  const { context, client, logger } = args;
+  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
+  const lastCallback = callbacks.pop();
+
+  if (lastCallback !== undefined) {
+    await processMiddleware(
+      callbacks, args, context, client, logger,
+      async () => lastCallback({ ...args, context, client, logger }),
+    );
+  }
+}
+
+export function isFunctionEvent(args: AnyMiddlewareArgs): args is AllWorkflowFunctionMiddlewareArgs {
+  return VALID_PAYLOAD_TYPES.has(args.payload.type);
+}
+
+function selectToken(context: Context): string | undefined {
+  // If attachFunctionToken = false, fallback to botToken or userToken
+  return context.functionBotToken ? context.functionBotToken : context.botToken || context.userToken;
```

Comment:
To @WilliamBergamin's earlier point, this area may need to be exposed as an API to the dev. Here we are making an assumption that for non-function-scoped events, the dev will want to use a bot token over a user token - that will not always be the case.

Reply:
This is how `WorkflowStep` was implemented, so merely a carry over. That was long ago, but is there a reality where an app has both present -- a bot token _and_ a user token? In my mind, I've always thought of it as one or the other (either it's a bot, or an app acting on behalf of the user), but you bringing this up makes me think that's a faulty mental model.

Reply:
Yes, an app can ask for scopes of both types (if you head to the OAuth & Permissions app config page, you can subscribe to both).

Just this week I was helping [someone who was using a user token for one API (uploading files) but a bot token in another (posting a message containing the uploaded file)](https://github.com/slackapi/bolt-python/issues/1016#issuecomment-1897096144).

Reply:
So, for any event that currently comes in, when we run `selectToken` when that event is processed, we opt for `botToken` first, followed by `userToken`. I think it's always been this way, or has for the last 4+ years.

@seratch your name is attached to one of those implementation lines. Can you advise here?

Reply:
I've always found this area to be poorly documented on our part and glossed over. For example bolt-js only mentions user scopes if you REALLY dig deep: you gotta scroll to the end of the OAuth docs, expand the hidden "Customizing OAuth options" section and then piece together the pieces based on the example and the one-line description of the `userScopes` field.

I believe, but may be wrong on this, that this has always been a userland concern that we don't provide guidance on. Like, for an app that is installable on-demand to workspace (implements OAuth) and collects both bot and user tokens, I'm pretty sure it is up to the installation store implementation (so userland code) to both: figure out which token to store as well as which token to select and return when doing the 'authorization' process that bolt exposes APIs for.

Reply:
Sorry, I overlooked this thread last week. The logic to select a primary token for context object from bot / user tokens is by design from the beginning, and all Bolt frameworks align.

We basically encourage developers to have a bot token, but technically it still should be allowed to go with only a user token (I don't think this is a common use case though). Thus, when a bot token is missing, setting a user token instead is a reassonable design for supporting such a scenario.

I don't think bolt should expose a way to customize this because:
1. we've never received such feature request in the last few years because most apps have their bot token
2. developers still can accesss both a bot and user token

As for the documentation topic, indeed this could be improved in the advanced topic section although the number of devs who need it is not so large.

Reply:
OK good to know and thanks for that context Kaz 🙇 
If Kaz says so then I defer to his suggestion / feel free to ignore my point